p2p/src/protocol.rs

```diff
@@ -15,256 +15,115 @@
+				Ok(None)
 			},
-			None,
-		)
-	}
-
-	/// Close the connection to the remote peer
-	fn close(&self) {
-		// TODO some kind of shutdown signal
-	}
-}
-
-impl ProtocolV1 {
-	fn send_msg<W: ser::Writeable>(&self, t: Type, body: &W) -> Result<(), Error> {
-		self.conn.borrow().send_msg(t, body)
-	}
 
-	fn send_request<W: ser::Writeable>(
-		&self,
-		t: Type,
-		rt: Type,
-		body: &W,
-		expect_resp: Option<Hash>,
-	) -> Result<(), Error> {
-		if self.conn.is_initialized() {
-			self.conn.borrow().send_request(t, rt, body, expect_resp)
-		} else {
-			Ok(())
-		}
-	}
-}
+			Type::Transaction => {
+				let tx: core::Transaction = msg.body()?;
+				adapter.transaction_received(tx);
+				Ok(None)
+			}
 
-fn handle_payload(
-	adapter: &NetAdapter,
-	sender: UnboundedSender<Vec<u8>>,
-	header: MsgHeader,
-	buf: Vec<u8>,
-	addr: SocketAddr,
-) -> Result<Option<Hash>, ser::Error> {
-	match header.msg_type {
-		Type::Ping => {
-			let ping = ser::deserialize::<Ping>(&mut &buf[..])?;
-			adapter.peer_difficulty(addr, ping.total_difficulty, ping.height);
-			let pong = Pong { total_difficulty: adapter.total_difficulty(), height: adapter.total_height() };
-			let mut body_data = vec![];
-			try!(ser::serialize(&mut body_data, &pong));
-			let mut data = vec![];
-			try!(ser::serialize(
-				&mut data,
-				&MsgHeader::new(Type::Pong, body_data.len() as u64),
-			));
-			data.append(&mut body_data);
+			Type::GetBlock => {
+				let h: Hash = msg.body()?;
+				debug!(LOGGER, "handle_payload: GetBlock {}", h);
 
-			if let Err(e) = sender.unbounded_send(data) {
-				debug!(LOGGER, "handle_payload: Ping, error sending: {:?}", e);
+				let bo = adapter.get_block(h);
+				if let Some(b) = bo {
+					let block_bytes = ser::ser_vec(&b).unwrap();
+					return Ok(Some((block_bytes, Type::Block)));
+				}
+				Ok(None)
 			}
 
-			Ok(None)
-		}
-		Type::Pong => {
-			let pong = ser::deserialize::<Pong>(&mut &buf[..])?;
-			adapter.peer_difficulty(addr, pong.total_difficulty, pong.height);
-			Ok(None)
-		},
-		Type::Transaction => {
-			let tx = ser::deserialize::<core::Transaction>(&mut &buf[..])?;
-			adapter.transaction_received(tx);
-			Ok(None)
-		}
-		Type::GetBlock => {
-			let h = ser::deserialize::<Hash>(&mut &buf[..])?;
-			debug!(LOGGER, "handle_payload: GetBlock {}", h);
+			Type::Block => {
+				let b: core::Block = msg.body()?;
+				let bh = b.hash();
 
-			let bo = adapter.get_block(h);
-			if let Some(b) = bo {
-				// serialize and send the block over
-				let mut body_data = vec![];
-				try!(ser::serialize(&mut body_data, &b));
-				let mut data = vec![];
-				try!(ser::serialize(
-					&mut data,
-					&MsgHeader::new(Type::Block, body_data.len() as u64),
-				));
-				data.append(&mut body_data);
-				if let Err(e) = sender.unbounded_send(data) {
-					debug!(LOGGER, "handle_payload: GetBlock, error sending: {:?}", e);
-				}
+				debug!(LOGGER, "handle_payload: Block {}", bh);
+
+				adapter.block_received(b, self.addr);
```

Comment:
This is where I'm wondering about multiple threads hitting the block processing pipeline in parallel. Two peer threads could receive the same block (or different blocks) simultaneously.

Reply:
Maybe worth considering having a cpupool/threadpool with a single thread that we can offload work to (and keep the peer threads for simply handling the network connections?)
If we haven't solved this another way?

Reply:
Correct, and it's solved by the sumtrees lock. I actually think it's a good thing because it allows us to start validating block headers in parallel.

Reply:
:+1: